File: Data Structure/Stack/arraystack.py
'''
Implement a stack using an array
'''

class ArrayStack:
    def __init__(self, init_size = 5):
        self.arr = [0.0 for _ in range(init_size)]
        self.top = -1
        self.n_element = 0

    def push(self, data):
        if self.n_element == len(self.arr):
            print("Increase capacity")
            self._increase_capacity()
        self.top += 1
        self.arr[self.top] = data
        self.n_element += 1

    def pop(self):
        if self.is_empty:
            print("Empty stack")
            return None
        pop = self.arr[self.top]
        self.top -= 1
        self.n_element -= 1
        return pop

    def size(self):
        return self.n_element

    def is_empty(self):
        if self.top < 0:
            return True
        else:
            return False

    def print_stack(self):
        if self.is_empty():
            print("Empty stack")
            return
        else:
            print(self.arr[0:self.top+1])

    def _increase_capacity(self):
        new_size = len(self.arr) * 2
        new_arr = [0.0 for _ in range(new_size)]
        for i, elem in enumerate(self.arr):
            new_arr[i] = elem
        self.arr = new_arr


if __name__ == "__main__":
    stk = ArrayStack(5)
    stk.push(2)
    stk.push(3)
    stk.push(4)
    stk.push(4)
    stk.print_stack()
    stk.push(5)
    stk.push(5)
    stk.print_stack()

File: Data Structure/Stack/array_stack.cpp
#include <iostream>
#include <stdio.h>

using namespace std;

class ArrayStack {
// Implement stack using arrary
public:
  int s[101];
  int top = -1;

  void push(int x) {
    cout << "Push in the stack" << endl;
    if ( top == 100 ) {
      cout << "Error: Over size!" << endl;
      return;
    }
    top++;
    s[top] = x;
    print_stack();
    cout << endl;
  }

  void pop() {
    cout << "Pop the top of the stack" << endl;
    if ( top == -1 ) {
      cout << "Error: No element!" << endl;
      return;
    }
    top--;
    print_stack();
    cout << endl;
  }

  void print_stack() {
    for (int i=0; i<=top; ++i) {
      cout << s[i] << ' ';
    }
    cout << endl;
  }
};//end class stack

int main(int argc, char const *argv[]) {
  ArrayStack astk;
  astk.push(1);
  astk.push(2);
  astk.push(3);
  astk.pop();
  astk.pop();
  return 0;
}
